dev-hash/src/hybrid.rs

```diff
@@ -0,0 +1,136 @@
+
+use std::hash::{Hash, Hasher};
+
+const FNV_OFFSET_BASIS: u64 = 0xcbf29ce484222325;
+const FNV_PRIME: u64 = 0x100000001b3;
+
+fn fnv_hash_bytes(bytes: &[u8]) -> u64 {
+    let mut hash = FNV_OFFSET_BASIS;
+    for &byte in bytes.iter() {
+        hash ^= byte as u64;
+        hash = hash.wrapping_mul(FNV_PRIME);
+    }
+    hash
+}
+
+fn fnv_hash_path<P: AsRef<Path>>(path: P) -> u64 {
+    let mut hasher = std::collections::hash_map::DefaultHasher::new();
+    path.as_ref().hash(&mut hasher);
+    let hash = hasher.finish();
+    fnv_hash_bytes(hash.to_le_bytes().as_slice())
+}
+
+#[derive(
+    Debug, Clone, PartialEq, Eq, Ord, PartialOrd, Hash, Serialize, Deserialize,
+)]
+pub struct Hybrid(pub String);
+
+impl FromStr for Hybrid {
+    type Err = hex::FromHexError;
+
+    fn from_str(s: &str) -> core::result::Result<Self, Self::Err> {
+        Ok(Hybrid(s.to_string()))
+    }
+}
+
+impl Display for Hybrid {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+const THRESHOLD: u64 = 1024 * 1024 * 1024;
+
+impl ResourceId for Hybrid {
+    fn from_path<P: AsRef<Path>>(file_path: P) -> Result<Self> {
+        let size = fs::metadata(file_path.as_ref())?.len();
+
+        if size < THRESHOLD {
+            // Use Blake3 for small files
+            log::debug!(
+                "Computing BLAKE3 hash for file: {:?}",
+                file_path.as_ref()
+            );
+
+            let file = fs::File::open(file_path)?;
+            let mut reader = BufReader::new(file);
+            let mut hasher = Blake3Hasher::new();
+            let mut buffer = Vec::new();
+            loop {
+                let bytes_read = reader.read_until(b'\n', &mut buffer)?;
+                if bytes_read == 0 {
+                    break;
+                }
+                hasher.update(&buffer);
+                buffer.clear();
+            }
+            let hash = hasher.finalize();
+            Ok(Hybrid(encode(hash.as_bytes())))
+        } else {
+            // Use fnv   hashing for large files
+            log::debug!(
+                "Computing simple hash for file: {:?}",
+                file_path.as_ref()
+            );
+
+            let hash = fnv_hash_path(file_path);
+            Ok(Hybrid(format!("{}_{}", size, hash)))
+        }
+    }
+
+    fn from_bytes(bytes: &[u8]) -> Result<Self> {
+        let size = bytes.len() as u64;
+        if size < THRESHOLD {
+            // Use Blake3 for small files
+            log::debug!("Computing BLAKE3 hash for bytes");
+
+            let mut hasher = Blake3Hasher::new();
+            hasher.update(bytes);
+            let hash = hasher.finalize();
+            Ok(Hybrid(encode(hash.as_bytes())))
+        } else {
+            // Use fnv hashing for large files
+            log::debug!("Computing simple hash for bytes");
+
+            let hash = fnv_hash_bytes(bytes);
+            Ok(Hybrid(format!("{}_{}", size, hash)))
```

Comment:
1. The original idea is the opposite: use Blake3 for small and medium files, and use faster function for large files where size of contents is large enough to make collision ratio low enough. 
2. FNV hashing can be added separately as dedicated hash function. Same as "skip-chunk" hash function.
3. A wild idea: can we parameterize this hybrid hash function with other hash functions? So we could compose 2 "dedicated" hash functions into threshold-based hash function.

Reply:
1. Yes, any file that has size below the `THRESHOLD` is being hashed by Blake3 already.
2. 100%
3. Yes, totally. I'm not sure if is there higher priority things to do before it, but we could even create a fully parameterized implementation, allowing indefinite pairs composed of a hash function and its related threshold. I've done something similar to it in JavaScript once

Reply:
1. Yes, any file that has size below the `THRESHOLD` is being hashed by Blake3 already.
2. 100%
3. Yes, totally. I'm not sure if there are higher priority things to do before it, but we could even create a fully parameterized implementation, that allows indefinite pairs composed of a hash function and its related threshold. I've done something similar to this in JavaScript once